types/src/unit_tests/trusted_state_test.rs

```diff
@@ -0,0 +1,521 @@
+                        let next_vset = into_validator_set(next_vset);
+                        let ledger_info = new_mock_ledger_info(epoch, version, Some(next_vset));
+                        let signatures = sign_ledger_info(&curr_vset[..], &ledger_info);
+
+                        epoch += 1;
+                        version += version_delta as u64;
+
+                        LedgerInfoWithSignatures::new(ledger_info, signatures)
+                    })
+                    .collect::<Vec<_>>();
+
+                // this will always succeed, since
+                // n >= 0, |vsets| = n + 1 ==> |vsets| >= 1
+                let last_vset = vsets.last().unwrap();
+                let latest_ledger_info = new_mock_ledger_info(epoch, version, None);
+                let signatures = sign_ledger_info(&last_vset[..], &latest_ledger_info);
+                let latest_ledger_info_with_sigs =
+                    LedgerInfoWithSignatures::new(latest_ledger_info, signatures);
+
+                (vsets, ledger_infos_with_sigs, latest_ledger_info_with_sigs)
+            })
+    })
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(100))]
+
+    #[test]
+    fn test_pre_genesis_trusted_state_always_ratchets_from_any_genesis(
+        (_vsets, lis_with_sigs, latest_li) in arb_update_proof(
+            0    /* start epoch */,
+            0    /* start version */,
+            1..3 /* version delta */,
+            1..3 /* epoch changes */,
+            1..5 /* validators per epoch */,
+        )
+    ) {
+        let trusted_state = TrustedState::new_trust_any_genesis_WARNING_UNSAFE();
+
+        let expected_latest_version = latest_li.ledger_info().version();
+        let expected_latest_li = latest_li.clone();
+        let expected_latest_epoch_change_li = lis_with_sigs.last().cloned();
+        let expected_validator_set = expected_latest_epoch_change_li
+            .as_ref()
+            .and_then(|li_with_sigs| li_with_sigs.ledger_info().next_validator_set());
+
+        let change_proof = ValidatorChangeProof::new(lis_with_sigs, false /* more */);
+        let trusted_state_change = trusted_state
+            .verify_and_ratchet(&latest_li, &change_proof)
+            .expect("Should never error or be stale when ratcheting from pre-genesis with valid proofs");
+
+        match trusted_state_change {
+            TrustedStateChange::Epoch {
+                new_state,
+                latest_li,
+                latest_epoch_change_li,
+                latest_validator_set,
+            } => {
+                assert_eq!(new_state.latest_version(), expected_latest_version);
+                assert_eq!(latest_li, &expected_latest_li);
+                assert_eq!(Some(latest_epoch_change_li), expected_latest_epoch_change_li.as_ref());
+                assert_eq!(Some(latest_validator_set), expected_validator_set);
+            }
+            TrustedStateChange::Version {
```

Comment:
Is this possible with epoch_changes=1..3?

Reply:
> I mean, a Err could be the server returning wrong number of responses. The msg is misleading.

ahh, good point. i'll just change it to unwrap then : )

Reply:
> Is this possible with epoch_changes=1..3?

also good catch. shouldn't be possible for a `Version` only change if there's always an epoch change ledger info generated